app/jobs/concerns/maintenance_tasks/task_job_concern.rb

```diff
@@ -157,10 +169,15 @@ def reenqueue_iteration_job(should_ignore: true)
     end
 
     def after_perform
+      new_status = @run.status
       @run.save!
       if defined?(@reenqueue_iteration_job) && @reenqueue_iteration_job
         reenqueue_iteration_job(should_ignore: false)
       end
+    rescue ActiveRecord::StaleObjectError
+      @run.reload_status
+      @run.status = new_status
```

Comment:
It feels a bit heavy-handed to force the new status here, is this to handle the running -> cancelling -> succeeded case which doesn't seem handled like the other ones in `on_complete`?
Like I mentioned I think there's still a race between on_shutdown and after_perform, so I think we should move the logic here. The hard part is that we want the transition logic from on_shutdown and on_complete to run, without the callbacks